install/0000_90_cluster-version-operator_02_servicemonitor.yaml

```diff
@@ -82,7 +82,11 @@ spec:
       annotations:
         message: Cluster operator {{ "{{ $labels.name }}" }} has been degraded for 10 minutes. Operator is degraded because {{ "{{ $labels.reason }}" }} and cluster upgrades will be unstable.
       expr: |
-        cluster_operator_conditions{job="cluster-version-operator", condition="Degraded"} == 1
+        (
+          cluster_operator_conditions{job="cluster-version-operator", condition="Degraded"}
+          or on (name)
+          group by (name) (cluster_operator_up{job="cluster-version-operator"})
```

Comment:
Having a hard time reading this:

"if operator is degraded OR is up == 1 for ten minutes report degraded?"

Shouldn't that be 

" if operator is degraded == 1 OR up == 0 for ten minutes report degraded"?

Reply:
I'm not sure what the OR portion is doing, so I'm not saying it's incorrect, but according to PR description ClusterOperatorDegraded should only fire when Degraded true or unset. And if up == 0 only ClusterOperatorDown should fire which is handled above.

Reply:
We have a bunch of ClusterOperators.  Grouping into sets:

a. Don't set `Available`, or set `Available!=True`.
b. Set `Available=True`.
c. Don't set `Degraded` or set it to an unrecognized value.
d. Set `Degraded=True`.
e. Set `Degraded=False`.

a∪b is all operators.  c∪d∪e is all operators.

`cluster_operator_up{job="cluster-version-operator"}` will be zero for (a) and one for (b).  `ClusterOperatorDown` will fire for (a) but not for (b).  Great.

`cluster_operator_conditions{job="cluster-version-operator", condition="Degraded"}` will be one for (d) and zero for (e), but misses (c).  `group by` collapses the `cluster_operator_up` value to one, regardless of (a) vs. (b).  So when those (c) clusters fall through to the `or`, they hit the `group by` one, and come out of the `(A or B)` block with one just like (d).  So `ClusterOperatorDegraded` will fire for (c) and (d) but not for (e).  Great.

Reply:
oh that's right.  i hate group by.